docs/src/markdown.ts

```diff
@@ -0,0 +1,266 @@
+      if (stats.isDirectory()) {
+        fileList = await getAllMdxFilePaths([filePath], fileList);
+      } else if (path.extname(file) === ".mdx") {
+        fileList.push(filePath);
+      }
+    }
+  }
+
+  return fileList;
+};
+
+// Returns the slugs of all headings in a tree
+const getHeadingsFromMarkdownTree = (
+  tree: ReturnType<typeof markdownProcessor.parse>
+): string[] => {
+  const headings: string[] = [];
+  slugger.reset();
+
+  visit(tree, "heading", (node) => {
+    let headingText = "";
+    // Account for headings with inline code blocks by concatenating the
+    // text values of all children of a heading node.
+    visit(node, (innerNode: any) => {
+      if (innerNode.value) {
+        headingText += innerNode.value;
+      }
+    });
+    const slugified = slugger.slug(headingText);
+    headings.push(slugified);
+  });
+
+  return headings;
+};
+
+/** Create a processor to parse MDX content */
+const markdownProcessor = unified()
+  .use(remarkParse)
+  .use(remarkRehype)
+  .use(rehypeRaw)
+  .use(function compiler() {
+    // A compiler is required, and we only need the AST, so we can
+    // just return it.
+    // @ts-ignore
+    this.Compiler = function treeCompiler(tree) {
+      return tree;
+    };
+  });
+
+const normalizePath = (filePath: string): string => {
+  const normalized = filePath
+    .replace("repo-docs", "/repo/docs")
+    .replace("pack-docs", "/pack/docs")
+    .replace(".mdx", "");
+
+  return normalized;
+};
+
+const validateFrontmatter = (path: string, data: Record<string, unknown>) => {
+  if (!data.title) {
+    throw new Error(`Document is missing a title: ${path}`);
+  }
+  if (!data.description) {
+    throw new Error(`Document is missing a description: ${path}`);
+  }
+  return data as {
+    title: string;
+    description: string;
+  };
+};
+
+/**
+ * Create a map of documents with their paths as keys and
+ * document content and metadata as values
+ * The key varies between doc pages and error pages
+ * error pages: `/docs/messages/example`
+ * doc pages: `api/example`
+ */
+const prepareDocumentMapEntry = async (
+  path: string
+): Promise<[string, Document]> => {
+  try {
+    const mdxContent = await fs.readFile(path, "utf8");
+    const { content, data } = matter(mdxContent);
+    const frontMatter = validateFrontmatter(path, data);
+
+    const tree = markdownProcessor.parse(content);
+    const headings = getHeadingsFromMarkdownTree(tree);
+    const normalizedUrlPath = normalizePath(path);
+
+    return [normalizedUrlPath, { content, path, headings, frontMatter }];
+  } catch (error) {
+    throw new Error(`Error preparing document map for file ${path}: ${error}`);
+  }
+};
+
+/** Checks if the links point to existing documents */
+const validateInternalLink =
+  (documentMap: Map<string, Document>) => (doc: Document, href: string) => {
+    // /docs/api/example#heading -> ["api/example", "heading""]
+    const [link, hash] = href.replace(DOCS_PATH, "").split("#", 2);
```

Comment:
Comment doesn't reflect the current `DOCS_PATH` value of `.`
```suggestion
    // /docs/api/example#heading -> ["/docs/api/example", "heading""]
    const [link, hash] = href.replace(DOCS_PATH, "").split("#", 2);
```